ourjapanlife/findadoc-server
# Rework schema, inputs, and tests

# What changed

- [x] Reviewed the whole gql schema
    - [x] removed `isDeleted`
    - [x] Removed unused queries for specialties, degrees, and locales. 
    - [x] added mutations for delete facility, professional, and submission
    - [x] Fixed a bug with optional arrays that should have been required
    - [x] Removed unnecessary nested entity input types (Ex. ContactInput, DegreeInput, etc.)

### Resolves #319, Resolves #320, Resolves #321
- [x] Created and reviewed add/update inputs for each entity. Ex. `addSubmissionInput` They should now only have appropriate fields for each action (add or update)
- [x] Updated all the resolvers and service functions to use updated schema.
    - [x] FacilityService
    - [x] HealthcareProfessionalService
    - [x] SubmissionService

### Resolves #322, Resolves #323
- [x] FIXED ALL THE TESTS! 🥳 
    - [x] Updated to use new input types.

### Resolves #278
- [x] Moved validation for `updateSubmission` from resolver into the submission service 

### Resolves #345
- [x] Simplified dbSchema and fixed it to match with gqlSchema (took a lot of work)

### Resolves #303
- [x] Created `updateHealthcareProfessional` 🥳 
    - [x] Create all the validation! 💯 

### Resolves #346
- [x] Reworked `createHealthcareProfessional`
    - [x] Also created validation for `createHealthcareProfessional` 💯 

### Resolves #347, Resolves #314
- [x] 💯 Replaced Jest with Vitest! 💯   Jest was causing many problems around ESM support, parallelization issues, and not giving clear error messages when tests were failing. Vitest uses the same exact naming as Jest so there was no tests migration really needed outside of the configurations. So far, it's significantly faster and more intuitive!
- [x] Upgraded our eslint, typescript, and node to use ESM (es modules). This caused a lot of issues and now everything is up to date. 

### Resolves #348
- [x] Centralized test setup and teadown in new `testSetup.test.ts` file. This fixed a parallelization problem with tests clearing other test data.  

### Resolves #349
- [x] Fixed our seed data! It is now all generated dynamically using the faker-js library (which is awesome)
    - [x]  Updated all the tests to use this as the input now instead of a local object. 

### General improvements
- [x] Added comments everywhere! 🗨️ 😄
- [x] Simplified the spoken language to just an enum (`languageCode_iso639_3`) instead of an object. The locale name is also updated to use the same enum. 
- [x] Unified all the `add` and `create` functions to use `create` naming
- [x] Added missing logic to `updateFacility` around managing sub-entities. 
- [x] Reworked `searchSubmissions` to use new `filters` input ✨ 
- [x] Converted CustomErrors in `healthcareProfessionalService` functions to use new `Result` object and proper validations ✨

### Bug fixes
- [x] added missing facility/healthcareprofessionals from submission (had to update everything to support this) 

# What is NOT included

- [ ] Created new tests
    - [ ] Created new test to check the facility/professional id associations work
    - [ ] Created new test to validate partial field updates work for Facility and healthcare provider. (that it will change only the fields sent and won't delete the ones that were not sent in)
    - [ ] Created `approveSubmission` tests 

- [ ] Created new logic for facility and healthcareproviders to keep their "foreign key" ids in sync when either is created or updated. (for example, a new healthcareprovider is created with a facilityid included, the facility should have its `healthcareProfessionalIds` updated with the new healthcareprovider id)

- Validation needs more love across all the services. Needs a separate ticket per service. 
- `searchHealthcareProfessionals` function needs to be created.
- a lot of our mapping function probably aren't needed, can be improved, or simplified to being just a validation function

NOTE TO SELF: let's move the language object to the front end and the key will be the language code enum.
`input SpokenLanguageInput {
  languageCode_iso639_3: languageCode_iso639_3!
  nameJa: String!
  nameEn: String!
  nameNative: String!
}
`

# Testing instructions

- Ensure the app builds and runs. `yarn dev`
- Run a test query like this [one](http://localhost:4000/?explorerURLState=N4IgJg9gxgrgtgUwHYBcQC4QEcYIE4CeABAIq6EBiAhlAJYA2tKtCAzgBQAkAZgyvq3RFqdRigIBlBFTxQAFhT4CAhAEoiwADpIiRbjQZMWHXvX55BRHkovqtO3UVphtjokiqIAokleOPiABSVH66ctJmclAyCAAKeBDcbKy0EB70AJJgrKFEUGkoNCgaubpUYGB4ySUObmUVVaysADK0SAgAjD6ljuWVya3tAEzdtXV9jS1tncE9unTis2NuCwSjdUQAvj0IcFQMPQDmEBCH9AgAslQADqwAqnj0PddyaQg9AO4IAEYp-LnbByAwEgAA0IAAbjJaFRvudWBgQPZdJoQKZzDkMDU3KjGHAmKihB1QblUYluKwEChCURiaSQAEEN0saiAFonVF%2BYFgkDhcoCRGckA81hQPC0a5oTAgTZAA)
- Ensure all tests run and pass `yarn test`
